API - Authentication chapter (#135)
infinum/rails-handbook: API/Authentication.md

_Note_: Cookie with the `HttpOnly` flag is the safest option for storing sensitive information in a browser, so if you have a JS-based frontend application that will consume your APIs, and you need to keep the secrets really secret, consider using this technique.

### Authorization header

HTTP supports sending a standard `Authorization` header in requests. The header consists of a type (indicating what kind of a credential it is) and a value separated by whitespace. An example header can look like this: `Basic aHR0cHM6Ly95b3V0dS5iZS9kUXc0dzlXZ1hjUQ==`.

[MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#authentication_schemes) has a nice list of types. The one most commonly seen in APIs is the so-called `Bearer` token. As explained in the [Bearer Token Usage RFC](https://datatracker.ietf.org/doc/html/rfc6750#section-1.2), a bearer token is a:

> A security token with the property that any party in possession of the token (a "bearer") can use the token in any way that any other party in possession of it can.
When the server generates a token, it should make it opaque (meaning they can't be interpreted by anyone other than the server).

> `Authorization` header is the preferred mechanism if your API is planned to be used by platforms other than the browser (browsers using it are susceptible to XSS attacks). It's easier to control than cookies outside of browsers, and it doesn't come with security concerns around using query strings (described below).

Reviewer: API for mobile, non-browser platforms should implement authorization header authentication.
API for browsers should implement cookies authentication.
API for both should ideally implement both, because cookies can complicate implementation on mobile platforms, and JS clients can't store/exchange auth information securely with the backend, other than through a cookie.

### Basic auth

> 

Reviewer: Since this is a type of Authorization mechanism, maybe we should indent it under Authorization chapter (by using 4 #, #### Basic auth).

# Authentication

Authentication is the process of verifying someone's identity. There are many ways to verify identities over HTTP APIs, the most popular of which will be described below.

When we talk about API authentication, it's important to distinguish its two components:
1. the mechanism — the medium which the authentication protocol utilizes
2. the method — the protocol for authenticating the user

These two work in tandem and sometimes they're indistinguishable, but we'll try to explain each one separately and then show how they work together. Let's look at available mechanisms first.

## Mechanisms

As was mentioned above, a mechanism is a medium utilized by the authentication protocol. The architecture of HTTP allows for several authentication mechanisms.

### Cookies

Cookies are pieces of data exchanged between the client and the server through headers. The server sends one or more cookies via a `Set-Cookie` response header and then the browser saves them and includes them in subsequent requests to the server in a `Cookie` header. The browser can also send cookies to the server without receiving any first. 

> When setting cookies, the server can apply one or more restrictions, like expiration date and domains where they're supposed to be used. You can learn more about how cookies work [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies).

Reviewer 2: ```suggestion
When setting cookies, the server can apply one or more restrictions, like expiration date and domains where they're supposed to be used. You can learn more about how cookies work on [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies).
```
? :)

There are two important cookie attributes you should know about:
- `Secure` — When set, cookie will only be sent to the server and saved in the browser if the protocol through which they're sent/received is `https`. Insecure `http` protocol can't access cookies with that attribute. If you're communicating over `https`, make sure to set this attribute.
- `HttpOnly` - These cookies cannot be accessed by JavaScript code. They are exchanged only between the browser and the server. If the cookie is used to store server-side session (a method described below), this attribute should be enabled because the cookie should be saved and used only by the browser, not read or changed by the user through JS. Note that this attribute doesn't prevent tampering of cookies as users can still edit them manually (for example, in a developer console).

> 

Reviewer 3: maybe mention that to the date, `Cookie + HttpOnly` is the safest option for storing sensitive information in a browser

Reviewer 4: Did you mean `Secure + HttpOnly` @vr4b4c?

Reviewer 4: done 👍

### Authorization header

HTTP supports sending a standard `Authorization` header in requests. The header consists of a type (indicating what kind of a credential it is) and a value separated by whitespace. An example header can look like this: `Basic aHR0cHM6Ly95b3V0dS5iZS9kUXc0dzlXZ1hjUQ==`.

> You can find the list of types [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#authentication_schemes). The one most commonly seen in APIs is the so-called `Bearer` token. As explained in the [Bearer Token Usage RFC](https://datatracker.ietf.org/doc/html/rfc6750#section-1.2), a bearer token is a:

Reviewer 2: ```suggestion
[MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#authentication_schemes) has a nice list of types. The one most commonly seen in APIs is the so-called `Bearer` token. As explained in the [Bearer Token Usage RFC](https://datatracker.ietf.org/doc/html/rfc6750#section-1.2), a bearer token is a:
```
This really is just a suggestion :)

> A security token with the property that any party in possession of the token (a "bearer") can use the token in any way that any other party in possession of it can.
When the server generates a token, it should make it opaque (meaning they can't be interpreted by anyone other than the server).

> `Authorization` header is the preferred mechanism if your API will be used in platforms other than the browser. It's easier to control than cookies outside of browsers, and it doesn't come with security concerns around using query strings (described below).

Reviewer 3: maybe mention it's still susceptible to the XSS attack

For example, URL `example.com?api_key=123&token=abc` has two query params: `api_key` with value `123`, and `token` with value `abc`.

Query string can be used to transmit keys and tokens in requests to the server, similar to cookies and `Authorization` header. At a first glance, this might seem insecure, but if it's sent over the `https` protocol, query string is encrypted with the rest of the request and it's visible only by the client and the server.

However, there are other concerns when using query strings to transmit sensitive data:
- URLs (including query strings) are visible in server logs (although, you can [filter parameters](https://guides.rubyonrails.org/action_controller_overview.html#parameters-filtering) like these in Rails)
- URLs are saved in browser history (when used in navigation)
- you have to include the key/token in every request, but normally you want to reserve the query string for other parameters (like filters, pagination, sorting, etc.)

> 

Reviewer 3: XSS attacks

> 

Reviewer 3: maybe mention length constraint

Consider using query strings for authentication only when other methods are unavailable.

## Methods

### API keys

> API keys are unique strings used to identify clients.

Reviewer 2: Think of API keys like normal passwords and you should treat them like passwords
